### PR TITLE
Add the permission backend.

### PR DESCRIPTION
**Pull request by Natim**
- [x] Write the base interface
- [x] Write the redis backend
- [x] Write the memory backend
- [x] Write permission backend generic unit tests
- [x] Write documentation

### CODE REVIEW

**Comment by Natim** [comment]
@leplatrem @ametaireau r?

File: cliquet/permission/__init__.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        :type object_id: string
+        :param permission: The permission object to remove the principal to.
+        :type permission: string
+        :returns: The list of user principals
+        :rtype: set
+
+        """
+        raise NotImplementedError
+
+    def has_permission(self, object_id, permission, user_id,
```
Comment (Natim) [inline cliquet/permission/__init__.py]: Replace ``user_id`` with ``principals``

File: cliquet/permission/__init__.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        """Remove a principal to a user.
+
+        :param user_id: The user_id to remove the principal to.
+        :type user_id: string
+        :param principal: The principal to remove to the user.
+        :type user_id: string
+        """
+        raise NotImplementedError
+
+    def get_user_principals(self, user_id):
```
Comment (Natim) [inline cliquet/permission/__init__.py]: Rename this ``get_user_groups``

File: cliquet/permission/__init__.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        :param str object_id: The object_id the permission is set to.
+        :param str permission: The permission object to remove the principal to.
+        :returns: The list of user principals
+        :rtype: set
+
+        """
+        raise NotImplementedError
+
+    def object_permission_authorized_principals(self, object_id, permission,
+                                                _get_perm_keys=None):
```
Comment (ametaireau) [inline cliquet/permission/__init__.py]: Do we want to have arguments starting with an `_`?

File: cliquet/permission/__init__.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        :type request: :class:`~pyramid:pyramid.request.Request`
+        :returns: ``True`` is everything is ok, ``False`` otherwise.
+        :rtype: bool
+        """
+        try:
+            self.add_user_principal(__HEARTBEAT_KEY__, 'alive')
+            self.remove_user_principal(__HEARTBEAT_KEY__, 'alive')
+        except BackendError:
+            return False
+        else:
```
Comment (ametaireau) [inline cliquet/permission/__init__.py]: superfluous else.

File: cliquet/permission/__init__.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        """Return the list of additionnal principals given to a user.
+
+        :param str user_id: The user_id to get the list of groups for.
+        :returns: The list of group principals the user is in.
+        :rtype: set
+
+        """
+        raise NotImplementedError
+
+    def add_object_permission_principal(self, object_id, permission,
```
Comment (ametaireau) [inline cliquet/permission/__init__.py]: `add_principal_to_ace`?

File: cliquet/permission/__init__.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                                        principal):
+        """Add a principal to an Access Control Entry.
+
+        :param str object_id: The object to add the permission principal to.
+        :param str permission: The permission object to add the principal to.
+        :param str principal: The principal to add to the ACE.
+        """
+        raise NotImplementedError
+
+    def remove_object_permission_principal(self, object_id, permission,
```
Comment (ametaireau) [inline cliquet/permission/__init__.py]: `remove_principal_from_ace`?

File: cliquet/permission/__init__.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+                                           principal):
+        """Remove a principal to an Access Control Entry.
+
+        :param str object_id: The object to remove the permission principal to.
+        :param str permission: The permission object to remove the principal to.
+        :param str principal: The principal to remove to the ACE.
+        """
+        raise NotImplementedError
+
+    def object_permission_principals(self, object_id, permission):
```
Comment (ametaireau) [inline cliquet/permission/__init__.py]: `get_required_principals`?

File: cliquet/permission/__init__.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+        :param str object_id: The object_id the permission is set to.
+        :param str permission: The permission object to remove the principal to.
+        :returns: The list of user principals
+        :rtype: set
+
+        """
+        raise NotImplementedError
+
+    def object_permission_authorized_principals(self, object_id, permission,
```
Comment (ametaireau) [inline cliquet/permission/__init__.py]: from the name it's hard to tell what's the difference with `object_permission_principals`.

Comment (Natim) [inline cliquet/permission/__init__.py]: The difference is inheritance, should we rename ``object_permission_inherited_principals``?

File: cliquet/permission/__init__.py (hunk 7 above)
Comment (Natim) [inline cliquet/permission/__init__.py]: I don't find this clearer, at least with the current name we know that it is ``object_permission_principals``, with ``get_required_principals`` we don't know what kind of principal we are talking about.

Also required let me think there are inherited but it is not the case here.

**Pull request closed without merging**